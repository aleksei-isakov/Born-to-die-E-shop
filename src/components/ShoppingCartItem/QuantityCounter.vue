<template>
  <div class="quantity__wrapper">
    <base-text-filled-button @click="onClickDecreaseQuantity"
      >-</base-text-filled-button
    >
    <input readonly :value="quantity" class="quantity__field" />
    <base-text-filled-button @click="onClickIncreaseQuantity"
      >+</base-text-filled-button
    >
  </div>
</template>

<script>
import { BaseTextFilledButton } from '@/base_components';

export default {
  name: 'QuantityCounter',

  components: { BaseTextFilledButton },

  props: {
    quantity: {
      type: Number,
      required: true,
      default: 1
    }
  },

  methods: {
    onClickIncreaseQuantity() {
      this.$emit('increase');
    },

    onClickDecreaseQuantity() {
      this.$emit('decrease');
    }
  }
};
</script>

<style lang="scss" scoped>
@import '@/scss/CustomVariables.scss';

.quantity__wrapper {
  display: flex;
}

.quantity__field {
  width: 60px;
  text-align: center;
  border: solid #e4e4e4ff;
  border-width: 2px 0;
}
</style>
