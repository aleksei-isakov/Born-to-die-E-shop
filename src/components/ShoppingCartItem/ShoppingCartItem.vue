<template>
  <div class="cart-item">
    <input
      type="checkbox"
      class="cart-item__checkbox"
      checked
      @change="onChangeToggleCartItemSelection($event.target.checked)"
    />
    <div class="cart-item__wrapper">
      <img :src="getImage" class="cart-item__img" />
      <div class="cart-item__name">{{ name }}</div>
      <div :class="{ 'cart-item__discount': discountPercentage }">
        {{ getDiscountPercentage }}
      </div>
      <div class="cart-item__price-container">
        <div v-if="discountPercentage" class="cart-item__price">
          {{ priceWithDiscount | currency }}
        </div>
        <div
          class="cart-item__price"
          :class="{ 'cart-item__crossed': discountPercentage }"
        >
          {{ price | currency }}
        </div>
      </div>
      <quantity-counter
        :quantity="quantity"
        @increase="increaseProductQuantity(id)"
        @decrease="decreaseProductQuantity(id)"
      />
    </div>
    <shopping-cart-trash-icon
      class="cart-item__trash-icon"
      @delete-item="deleteFromCart(id)"
    />
  </div>
</template>

<script>
import defaultImage from '@/assets/defaultImage.jpg';
import QuantityCounter from '@/components/ShoppingCartItem/QuantityCounter';
import ShoppingCartTrashIcon from '@/components/ShoppingCartTrashIcon/ShoppingCartTrashIcon.vue';
import { mapActions } from 'vuex';

export default {
  name: 'ShoppingCartItem',

  components: { QuantityCounter, ShoppingCartTrashIcon },

  props: {
    images: {
      type: Array,
      default: () => []
    },

    id: {
      type: String,
      required: true,
      default: ''
    },

    name: {
      type: String,
      required: true,
      default: ''
    },

    quantity: {
      type: Number,
      required: true,
      default: 1
    },

    price: {
      type: Number,
      required: true,
      default: 0
    },

    discountPercentage: {
      type: Number,
      required: true,
      default: 0
    },

    priceWithDiscount: {
      type: Number,
      required: true,
      default: 0
    }
  },

  computed: {
    getImage() {
      if (this.images && this.images.length) {
        return this.images[0];
      }

      return defaultImage;
    },

    // getPrice() {
    //   return `${this.price} $`;
    // },

    // getPriceWithDiscount() {
    //   return `${this.priceWithDiscount.toFixed(1)} $`;
    // },

    getDiscountPercentage() {
      return this.discountPercentage > 0 ? `${this.discountPercentage}%` : '';
    }
  },

  methods: {
    ...mapActions('ShoppingCartModule', [
      'increaseProductQuantity',
      'decreaseProductQuantity',
      'deleteFromCart',
      'checkCartItem',
      'uncheckCartItem'
    ]),

    onChangeToggleCartItemSelection(checked) {
      if (checked) {
        return this.checkCartItem(this.id);
      }

      return this.uncheckCartItem(this.id);
    }
  }
};
</script>

<style lang="scss" scoped>
@import '@/scss/CustomVariables.scss';

.cart-item {
  position: relative;
  display: flex;
  gap: 10px;
  width: 100%;
  align-items: center;

  &__checkbox {
    flex-shrink: 0;
    width: 20px;
    height: 20px;
    margin-right: 5px;
    cursor: pointer;
  }

  &__wrapper {
    display: grid;
    align-items: center;
    grid-template-columns: 1fr 1fr 0.5fr 1fr 1fr;
    flex-grow: 1;
    gap: 20px;
    min-width: 260px;
    padding: 10px 65px 10px 10px;
    color: $font-color-text;
    border: 1px solid #e4e4e4ff;
    border-radius: 4px;
    box-shadow: $shadow;
    background-color: $white;
  }

  &__name {
    text-align: left;
  }

  &__img {
    align-content: flex-start;
    max-width: 130px;
  }
  &__price-container {
    display: flex;
    align-items: center;
  }

  &__price {
    min-width: fit-content;
    font-size: $font-size-subtitle;
    color: $font-color-subtitle;
    text-align: left;
  }

  &__trash-icon {
    position: absolute;
    right: 20px;
  }

  &__crossed {
    text-decoration: line-through;
    color: $font-color-subtitle;
    font-size: 14px;
    margin-left: 10px;
  }

  &__discount {
    color: $white;
    background-color: #b61d1c;
    padding: 6px 16px;
    border-radius: 20px;
    width: fit-content;
  }
}

@media screen and (max-width: $tablet-size) {
  .cart-item {
    justify-content: center;

    &__wrapper {
      grid-template-columns: 1fr;
      justify-items: center;
      max-width: 360px;
      padding: 20px;
    }

    &__img {
      max-width: 100%;
    }

    &__trash-icon {
      position: static;
    }
  }
}
</style>
