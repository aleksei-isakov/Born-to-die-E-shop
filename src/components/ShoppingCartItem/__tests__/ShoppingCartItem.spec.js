import Vue from 'vue';
import Vuetify from 'vuetify';
import { mount, shallowMount, createLocalVue } from '@vue/test-utils';
import ShoppingCartItem from '@/components/ShoppingCartItem/ShoppingCartItem';
import defaultImage from '@/assets/defaultImage.jpg';
import Vuex from 'vuex';
Vue.use(Vuetify);

describe('ShoppingCartItem', () => {
  let wrapper;
  const name = 'name';
  const price = 35;
  const discountPercentage = 0;
  const priceWithDiscount = 35;
  const quantity = 1;
  const id = '0';
  let vuetify = new Vuetify();
  const imageList = [
    'http://placeimg.com/640/480',
    'http://placeimg.com/640/480',
    'http://placeimg.com/640/480'
  ];

  beforeEach(() => {
    wrapper = mount(ShoppingCartItem, {
      vuetify,
      propsData: {
        images: imageList,
        id: id,
        name: name,
        quantity: quantity,
        price: price,
        priceWithDiscount: priceWithDiscount,
        discountPercentage: discountPercentage
      }
    });
  });

  afterEach(() => {
    wrapper.destroy();
  });

  test('is a Vue instance', () => {
    expect(wrapper.isVueInstance).toBeTruthy();
  });

  test('should return correct props', () => {
    expect(wrapper.props().name).toBe(name);
    expect(wrapper.props().price).toBe(price);
    expect(wrapper.props().quantity).toBe(quantity);
    expect(wrapper.props().images).toBe(imageList);
    expect(wrapper.props().id).toBe(id);
  });

  test('Should take the first image if images array is not empty', () => {
    const localThis = { images: imageList };

    expect(ShoppingCartItem.computed.getImage.call(localThis)).toBe(
      imageList[0]
    );
  });

  test('Should get default image when no images in the props', () => {
    const localThis = { images: null };

    expect(ShoppingCartItem.computed.getImage.call(localThis)).toBe(
      defaultImage
    );
  });

  test('match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});

const localVue = createLocalVue();

localVue.use(Vuex);

describe('Checkbox actions', () => {
  let actions;
  let store;

  beforeEach(() => {
    actions = {
      checkCartItem: jest.fn(),
      uncheckCartItem: jest.fn()
    };

    store = new Vuex.Store({
      modules: {
        ShoppingCartModule: {
          namespaced: true,
          state: {
            productsInCart: []
          },
          actions
        }
      }
    });
  });

  it('dispatches "toggleCartItemSelection" when checkbox is checked', () => {
    const wrapper = shallowMount(ShoppingCartItem, {
      store,
      localVue,
      propsData: {
        id: '1',
        price: 1,
        quantity: 1,
        name: 'name',
        discountPercentage: 0,
        priceWithDiscount: 1
      }
    });

    wrapper.vm.onChangeToggleCartItemSelection(true);
    expect(actions.checkCartItem).toHaveBeenCalled();

    wrapper.vm.onChangeToggleCartItemSelection(false);
    expect(actions.uncheckCartItem).toHaveBeenCalled();
  });

  it('returns price with discount', () => {
    const wrapper = shallowMount(ShoppingCartItem, {
      store,
      localVue,
      propsData: {
        id: '1',
        price: 100,
        quantity: 1,
        name: 'name',
        discountPercentage: 33.3333,
        priceWithDiscount: 66.66667
      }
    });

    expect(wrapper.vm.getPriceWithDiscount).toEqual('66.7 $');
  });

  it('returns discount percentage', () => {
    const wrapper = shallowMount(ShoppingCartItem, {
      store,
      localVue,
      propsData: {
        id: '1',
        price: 100,
        quantity: 1,
        name: 'name',
        discountPercentage: 50,
        priceWithDiscount: 50
      }
    });

    expect(wrapper.vm.getDiscountPercentage).toEqual('50%');
  });
});
