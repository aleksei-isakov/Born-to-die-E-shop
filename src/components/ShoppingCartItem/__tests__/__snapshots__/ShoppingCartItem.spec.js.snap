// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShoppingCartItem match snapshot 1`] = `
<div class="cart-item"><input type="checkbox" checked="checked" class="cart-item__checkbox">
  <div class="cart-item__wrapper"><img src="http://placeimg.com/640/480" class="cart-item__img">
    <div class="cart-item__name">name</div>
    <div class="cart-item__price">35 $</div>
    <div class="quantity__wrapper"><button class="base-button base-text-button base-text-filled-button custom-blue">-</button> <input readonly="readonly" class="quantity__field"> <button class="base-button base-text-button base-text-filled-button custom-blue">+</button></div>
    <div class="title-icon trash-icon"><img src="" width="25" alt="icon"></div>
  </div>
</div>
`;
